graph/push_v2.go

```diff
@@ -0,0 +1,251 @@
+			break
+		}
+
+		logrus.Debugf("Pushing layer: %s", layer.ID)
+
+		if layer.Config != nil && metadata.Image != layer.ID {
+			if err := runconfig.Merge(&metadata, layer.Config); err != nil {
+				return err
+			}
+		}
+
+		jsonData, err := p.graph.RawJSON(layer.ID)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve the path for %s: %s", layer.ID, err)
+		}
+
+		// Push v1 compatibility object for compatibility registry
+		// Set jsonData and layer.ID
+
+		var exists bool
+		dgst, err := p.graph.GetDigest(layer.ID)
+		switch err {
+		case nil:
+			_, err := repo.Blobs(nil).Stat(nil, dgst)
+			switch err {
+			case nil:
+				exists = true
+				out.Write(sf.FormatProgress(stringid.TruncateID(layer.ID), "Image already exists", nil))
+			case distribution.ErrBlobUnknown:
+				// nop
+			default:
+				out.Write(sf.FormatProgress(stringid.TruncateID(layer.ID), "Image push failed", nil))
+				return err
+			}
+		case ErrDigestNotSet:
+			// nop
+		case digest.ErrDigestInvalidFormat, digest.ErrDigestUnsupported:
+			return fmt.Errorf("error getting image checksum: %v", err)
+		}
+
+		// if digest was empty or not saved, or if blob does not exist on the remote repository,
+		// then fetch it.
+		if !exists {
+			if pushDigest, err := p.pushV2Image(repo.Blobs(nil), layer, sf, out); err != nil {
+				return err
+			} else if pushDigest != dgst {
+				// Cache new checksum
+				if err := p.graph.SetDigest(layer.ID, pushDigest); err != nil {
+					return err
+				}
+				dgst = pushDigest
+			}
+		}
+
+		m.FSLayers = append(m.FSLayers, manifest.FSLayer{BlobSum: dgst})
+		m.History = append(m.History, manifest.History{V1Compatibility: string(jsonData)})
+
+		layersSeen[layer.ID] = true
+	}
+
+	logrus.Infof("Signed manifest for %s:%s using daemon's key: %s", repo.Name(), tag, p.trustKey.KeyID())
+	signed, err := manifest.Sign(m, p.trustKey)
+	if err != nil {
+		return err
+	}
+
+	manifestDigest, err := digestFromManifest(signed, repo.Name())
+	if err != nil {
+		return err
+	}
+	if manifestDigest != "" {
+		out.Write(sf.FormatStatus("", "Digest: %s", manifestDigest))
+	}
+
+	return repo.Manifests().Put(signed)
+}
+
+func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *Image, sf *streamformatter.StreamFormatter, out io.Writer) (digest.Digest, error) {
+	out.Write(sf.FormatProgress(stringid.TruncateID(img.ID), "Buffering to Disk", nil))
+
+	image, err := p.graph.Get(img.ID)
+	if err != nil {
+		return "", err
+	}
+	arch, err := p.graph.TarLayer(image)
+	if err != nil {
+		return "", err
+	}
+
+	tf, err := p.graph.newTempFile()
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		tf.Close()
+		os.Remove(tf.Name())
+	}()
+
+	size, dgst, err := bufferToFile(tf, arch)
+	if err != nil {
+		return "", err
+	}
+
+	// Send the layer
+	logrus.Debugf("rendered layer for %s of [%d] size", img.ID, size)
+	layerUpload, err := bs.Create(nil)
+	if err != nil {
+		return "", err
+	}
+	// TODO(tiborvass): @dmcgowan need review
```

Comment:
A `BlobWriter` should always be closed regardless of whether `Commit` or `Cancel` is called. This comment can be removed.